ZenPacks/zenoss/HttpMonitor/http.py

```diff
@@ -10,107 +10,113 @@
+    def _getIp(self, response):
+        if response:
+            for a in response[0]:
+                if a.payload.TYPE == dns.A:
+                    self._hostnameIp.append(a.payload.dottedQuad())
+        return self.request()
 
-    def seturl(self, url):
+    def setProp(self, ipAddr, hostname, url="/", port=80, timeout=5, ssl=False, follow=True):
+        self._ipAddr = ipAddr
+        self._port = port
+        self._hostname = hostname
         self._url = url
-        HTTPClientFactory.__init__(self, url=self._url, followRedirect=self._follow)
-        self.setURL(self._url)
-        self._factory = self
-        return self
-
-    def setURL(self, url):
-        HTTPClientFactory.setURL(self, url)
-        self.path = url
-
-    def page(self, page):
-        if self.waiting:
-            self.waiting = 0
-            res = dict()
-            res['body'] = page
-            res['headers'] = self.response_headers
-            res['code'] = self.status
-            res['message'] = self.message
-            res['time'] = time.time() - self._startTime
-            res['size'] = self._bodysize(page)
-            self.deferred.callback(res)
-
-    def noPage(self, reason):
-        if self.waiting:
-            self.waiting = 0
-            self.deferred.errback(reason)
-
-    def _bodysize(self, body=""):
-        return sys.getsizeof(body)
+        self._timeout = timeout
+        self._ssl = ssl
+        self._reactor = reactor
+        self._clientEdnpoint = None
+        self._proxyIp = None
+        self._reqURL = ""
+        self._startTime = None
+        self._follow = follow
+        self._response = dict()
+        self._headers = Headers({b"User-Agent": [b"Zenoss HttpMonitor"]})
+        self._body = ""
+        self._hostnameIp = list()
+
+    def makeURL(self):
+        url_data = URI.fromBytes(self._url)
+        if not url_data.scheme:
+            scheme = "https" if self._ssl else "http"
+            self._reqURL = '{scheme}://{hostname}:{port}{url}'.format(scheme=scheme, hostname=self._hostname,
+                                                                      port=self._port, url=self._url)
+            return
+        else:
+            if url_data.host == self._hostname and self._ipAddr in self._hostnameIp:
+                scheme = "https" if self._ssl else "http"
+                self._reqURL = '{scheme}://{hostname}:{port}{url}'.format(scheme=scheme, hostname=self._hostname,
+                                                                          port=self._port, url=url_data.path)
+                return
+            else:
+                self._proxyIp = self._ipAddr
+                self._reqURL = '{scheme}://{hostname}:{port}{url}'.format(scheme=url_data.scheme,
+                                                                          hostname=url_data.host, port=url_data.port,
+                                                                          url=url_data.path)
+                return
```

Comment:
This `makeURL` can be simplified.  I'll use a truth table exercise to determine what's necessary.

The inputs are:
1. URL has a scheme
2. URL's host is given hostname
3. Given IP address is the hostname's IP address

And the outputs are:
1. scheme
2. host
3. port
4. path

 The first truth table maps the inputs to outputs in every permutation *as written in your PR*:
```
 -------------------------------------------------------------------------------
 URL has | URL host    | ipAddr is ||
 scheme  | is hostname | host IP   ||  scheme  |   host   |   port   |   path
 --------+-------------+-----------++----------+----------+----------+----------
  True   |    True     |   True    ||   ssl    |   host   |   port   | from url
  True   |    True     |   False   || from url | from url | from url | from url
  True   |    False    |   True    || from url | from url | from url | from url
  True   |    False    |   False   || from url | from url | from url | from url
  False  |    True     |   True    ||   ssl    |   host   |   port   |   url
  False  |    True     |   False   ||   ssl    |   host   |   port   |   url
  False  |    False    |   True    ||   ssl    |   host   |   port   |   url
  False  |    False    |   False   ||   ssl    |   host   |   port   |   url
 -------------------------------------------------------------------------------
```
So immediately, we can see there's some redundancy, so we can reduce this table to one line when "URL has scheme" is False:
```
 -------------------------------------------------------------------------------
 URL has | URL host    | ipAddr is ||
 scheme  | is hostname | host IP   ||  scheme  |   host   |   port   |   path
 --------+-------------+-----------++----------+----------+----------+----------
  True   |    True     |   True    ||   ssl    |   host   |   port   | from url
  True   |    True     |   False   || from url | from url | from url | from url
  True   |    False    |   True    || from url | from url | from url | from url
  True   |    False    |   False   || from url | from url | from url | from url
  False  |     any     |    any    ||   ssl    |   host   |   port   |   url
 -------------------------------------------------------------------------------
```
There's redundancy when "URL host is hostname" is False, so another reduction:
```
 -------------------------------------------------------------------------------
 URL has | URL host    | ipAddr is ||
 scheme  | is hostname | host IP   ||  scheme  |   host   |   port   |   path
 --------+-------------+-----------++----------+----------+----------+----------
  True   |    True     |   True    ||   ssl    |   host   |   port   | from url
  True   |    True     |   False   || from url | from url | from url | from url
  True   |    False    |    any    || from url | from url | from url | from url
  False  |     any     |    any    ||   ssl    |   host   |   port   |   url
 -------------------------------------------------------------------------------
```
Analyzing the first line in the table reveals that regardless of whether "ipAddr is host IP", if the URL's host and the given hostname are the same, the request URL should be the URL's host (which is the same as hostname).  This allows us to reduce the table to two lines:
```
 -------------------------------------------------------------------------------
 URL has | URL host    | ipAddr is ||
 scheme  | is hostname | host IP   ||  scheme  |   host   |   port   |   path
 --------+-------------+-----------++----------+----------+----------+----------
  True   |     any     |    any    || from url | from url | from url | from url
  False  |     any     |    any    ||   ssl    |   host   |   port   |   url
 -------------------------------------------------------------------------------
```
With this information, creating the request URL becomes pretty simple:
```
url_data = URI.fromBytes(self._url)
if url_data.scheme:
    args = {
        "scheme": url_data.scheme,
        "host": url_data.host,
        "port": url_data.port,
        "path": url_data.path,
    }
else:
    args = {
        "scheme": "https" if self._ssl else "http",
        "hostname": self._hostname,
        "port": self._port,
        "path": self._url,
    }
self._reqUrl = "{scheme}://{hostname}:{port}{path}".format(**args)
```
Now, what to do with `_proxyIp`?  Let's use another truth table.
```
 ----------------------------------------------
 URL has | URL host    | ipAddr is ||
 scheme  | is hostname | host IP   || proxyIp
 --------+-------------+-----------++----------
  True   |    True     |   True    ||  None
  True   |    True     |   False   ||  ipAddr
  True   |    False    |   True    ||  ipAddr
  True   |    False    |   False   ||  error
  False  |     n/a     |   True    ||  None
  False  |     n/a     |   False   ||  ipAddr
  False  |     n/a     |   True    ||  None
  False  |     n/a     |   False   ||  ipAddr
 ----------------------------------------------
```
Simple reduction when "URL has scheme" is False:
```
 ----------------------------------------------
 URL has | URL host    | ipAddr is ||
 scheme  | is hostname | host IP   || proxyIp
 --------+-------------+-----------++----------
  True   |    True     |   True    ||  None
  True   |    True     |   False   ||  ipAddr
  True   |    False    |   True    ||  ipAddr
  True   |    False    |   False   ||  error
  False  |     n/a     |   True    ||  None
  False  |     n/a     |   False   ||  ipAddr
 ----------------------------------------------
```
This is more complicated and the reason why I contemplated that perhaps the datasource parameters should be refactored into something more obvious (and reasonable). The code to implement this table looks like:
```
from operator import xor
...
url_data = URI.fromBytes(self._url)
hasScheme = bool(url_data.scheme)
hostMatch = url_data.host == self._hostname
ipMatch = self._ipAddr in self._hostnameIp
if (hasScheme and not hostMatch) or \
        (not hasScheme and xor(hostMatch, ipMatch)):
    self._proxyIp = self._ipAddr
```
I ignored the error condition.